OGM-983. changes from EPAM.

schernolyas opened the pull request:
change list:
1) migration to Ignite 2.1
2) support automatic schema generation for Ignite
3) support Native Ignite Persistence
4) support Ignite transactions by transaction emulation

Commit:
OGM-983. changes from EPAM.

list:
1) migration to Ignite 2.1
2) support automatic schema generation for Ignite
3) support Native Ignite Persistence
4) support Ignite transctions by transaction emulation

Commit:
OGM-983. correct order of imports

Commit:
OGM-983. remove unused changes

Commit:
OGM-983. remove unused changes

schernolyas committed:
OGM-983.  Add method for getting currect instance of Ignite

schernolyas committed:
OGM-983 add serialVersionUID

DavideD commented:
Jenkins, retest this please

DavideD commented on ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteDatastoreProvider.java:
```diff
@@ -328,6 +328,7 @@ public void clearCache(String cacheName) {
 	 */
 	private static class ComputeForLocalQueries<T> implements IgniteCallable<List<T>> {
 
+		private static final long serialVersionUID = 1L;
```
We don't add the serial version id to classes unless there are good reason to have it.

schernolyas replied:
ok. I will remove it. Eclipse asks

DavideD replied:
I know but it is a warning. Let it ask :)

schernolyas replied:
fixed

DavideD commented on ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteProperties.java:
```diff
@@ -29,6 +32,36 @@
 	 */
 	public static final String IGNITE_INSTANCE_NAME = "hibernate.ogm.ignite.instance_name";
 
+	public static final String IGNITE_ALLOWS_TRANSACTION_EMULATION = "hibernate.ogm.ignite.allows_transaction_emulation";
```
By the way,  this property is ignored by OGM if you are using a custom `TransactionCoordinatorBuilder`. Please remove it

schernolyas committed:
OGM-983. remove unused serialVersionUID field

DavideD commented on ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteCacheInitializer.java:
```diff
@@ -33,9 +43,20 @@
+		h2TypeMapping = Collections.unmodifiableMap( map );
+	}
+
+
```
Could explain me what is it this h2TypeMapping?

schernolyas replied:
it is type changing.
returnValue = h2TypeMapping.getOrDefault( returnValue , returnValue  );
queryEntity.addQueryField( currentColumn.getName(),returnValue.getName(),null );

Reason: h2 not have field type 'character' and I use the changing for fix BuiltInTypeTest.testCharacterSupport()

viktor44 commented on ignite/src/main/java/org/hibernate/ogm/datastore/ignite/transaction/impl/IgniteTransactionCoordinatorBuilder.java:
```diff
@@ -0,0 +1,62 @@
+		if ( delegate.isJta() ) {
+			//return new IgniteJtaTransactionCoordinator( coordinator, datastoreProvider );
+			//@todo add currect code of Victor
+			throw new UnsupportedOperationException( "Not supported yet" );
```
Do you remove support of JTA?

schernolyas replied:
Hi @Z-z-z-z !
No.  JTA supports by your code. And ... it is main question. Can I migrate your code to IgniteJtaTransctionCoordinator like in all other dialects?

DavideD replied:
@schernolyas I think we agreed that before changing/applying any of this you need to provide a test case.

viktor44 commented:
org.hibernate.ogm.backendtck.associations.collection.unidirectional.CollectionUnidirectionalTest_
org.hibernate.ogm.backendtck.associations.collection.manytomany.ManyToManyExtraTest#testUnidirectionalManyToMany
org.hibernate.ogm.backendtck.associations.manytoone.ManyToOneTest#testUnidirectionalManyToOne
org.hibernate.ogm.backendtck.associations.manytoone.ManyToOneTest#testRemovalOfTransientEntityWithAssociation
org.hibernate.ogm.backendtck.jpa.JPAJTATest
org.hibernate.ogm.backendtck.jpa.JPAResourceLocalTest
org.hibernate.ogm.backendtck.id.SequenceIdGeneratorTest
org.hibernate.ogm.backendtck.id.SequenceNextValueGenerationTest
org.hibernate.ogm.backendtck.compensation.CompensationSpiJpaTest


these tests wasn't skipped before

schernolyas commented:
Okey ... I will fix

On 16 September 2017 at 02:30, Victor <notifications@github.com> wrote:

> org.hibernate.ogm.backendtck.associations.collection.unidirectional.
> CollectionUnidirectionalTest_
> org.hibernate.ogm.backendtck.associations.collection.manytomany.
> ManyToManyExtraTest#testUnidirectionalManyToMany
> org.hibernate.ogm.backendtck.associations.manytoone.ManyToOneTest#
> testUnidirectionalManyToOne
> org.hibernate.ogm.backendtck.associations.manytoone.ManyToOneTest#
> testRemovalOfTransientEntityWithAssociation
> org.hibernate.ogm.backendtck.jpa.JPAJTATest
> org.hibernate.ogm.backendtck.jpa.JPAResourceLocalTest
> org.hibernate.ogm.backendtck.id.SequenceIdGeneratorTest
> org.hibernate.ogm.backendtck.id.SequenceNextValueGenerationTest
> org.hibernate.ogm.backendtck.compensation.CompensationSpiJpaTest
>
> these tests wasn't skipped before
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hibernate/hibernate-ogm-ignite/pull/3#issuecomment-329926309>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI4fSxxIeplVPCiUEH0fNgz47UHqmvqLks5siwiVgaJpZM4PXNF7>
> .
>



-- 
---------------------

With best regards, Sergey Chernolyas

schernolyas committed:
OGM-983. fix test PolymorphicAssociationTest

test
'org.hibernate.ogm.backendtck.inheritance.singletable.depositor.PolymorphicAssociationTest'
fixed

schernolyas committed:
OGM-983 fix test SequenceIdGeneratorTest

fix test org.hibernate.ogm.backendtck.id.SequenceIdGeneratorTest

schernolyas committed:
OGM-983 exclude test

exclude org.hibernate.ogm.backendtck.id.SequenceNextValueGenerationTest
from list of skiped tests.

viktor44 commented on ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteProperties.java:
```diff
@@ -29,6 +32,36 @@
+	 * Configuration property 'work.directory.path' set path for work directory
+	 * @see <a href="https://apacheignite.readme.io/docs/distributed-persistent-store">Ignite Native Persistence</a>
+	 */
+	public static final String IGNITE_WORK_DIRECTORY = "hibernate.ogm.ignite.work.directory.path";
```
Why did you do IGNITE_SUPPORT_PERSISTENCE and IGNITE_WORK_DIRECTORY properties?
They can be set trought CONFIGURATION_CLASS_NAME. There are many other properties in IgniteConfiguration. I don't think that work directory is something special

schernolyas replied:
For activate Ignite Native Persistence, Hibenate needs call "ignite.active(true);". You right that these properties can be set by configuration file/class. Any way, You need to send command to Hibernate to call "ignite.active(true);" after Ignite started.
In this way, property IGNITE_WORK_DIRECTORY can be removed from the class, but IGNITE_SUPPORT_PERSISTENCE is not.
Correct me please if I make mistake.

viktor44 replied:
IMHO trying to start real cluster, especially with local storage from Hibernate configuration is a big mistake. Better ways are to use OGM in Ignite client mode or to start cluster before Hibernate and then use IGNITE_INSTANCE_NAME to work with Ignite cluster. 
But if you still want to call active(true) you can do it always when clientMode=false or if persistentStoreConfiguration != null and clientMode=false.

schernolyas replied:
okey. Good idea!

viktor44 commented on ignite/src/main/java/org/hibernate/ogm/datastore/ignite/IgniteDialect.java:
```diff
@@ -146,12 +145,13 @@ public void insertOrUpdateTuple(EntityKey key, TuplePointer tuplePointer, TupleC
-			builder = provider.createBinaryObjectBuilder( tupleSnapshot.getCacheValue() );
+			IgniteTupleSnapshot tupleSnapshot = (IgniteTupleSnapshot) tuple.getSnapshot();
+			keyObject = tupleSnapshot.getCacheKey();
+			builder = provider.createBinaryObjectBuilder( entityCache.get( keyObject ) );
```
I think that you replace my new code with old vesrion. What was wrong in insertOrUpdateTuple?
In any case `entityCache.get( keyObject )` here isn't good.

schernolyas replied:
May be.... Where can I take a last version of your code for merging?

schernolyas replied:
Do you mean this commit 'https://github.com/Z-z-z-z/hibernate-ogm-ignite/commit/4835b9718b95748623910fe5008b72aee7ed025f'?

viktor44 replied:
yes

schernolyas committed:
OGM-983 correct cache key generation from EntityKey

schernolyas committed:
OGM-983. unskip tests

schernolyas committed:
OGM-983. change default concurency

DavideD commented:
@schernolyas I think this PR is too complicated and I can't merge it if you don't split it in smaller chunks. The problem is that the main commit contains many different changes unrelated among each other. Some of the changes make sense but others require more work or there is not a consensus about merging them. 

I would like to see each of the elements in the change list in its own PR containing the minimum set of changes to make it work:

1. Migration to Ignite 2.1
2. Support Automatic schema generation
3. Additional properties (if you need them)

The current way of dealing transaction seems the correct one to me, but if you want to change it and use the transaction emulation approach you should do it in a separate PR and adding a test case that show
the problem you are trying to solve.  Or at least try to create a test project that I can run to understand why you need it.

DavideD commented:
I'm going to close this PR because @schernolyas agreed to split it in several different new PRs
